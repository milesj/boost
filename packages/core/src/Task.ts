/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import Context from './Context';
import {
  STATUS_PENDING,
  STATUS_RUNNING,
  STATUS_SKIPPED,
  STATUS_PASSED,
  STATUS_FAILED,
} from './constants';
import { Status } from './types';
import Emitter from './Emitter';

export type TaskAction<Ctx extends Context> = (
  context: Ctx,
  value: any,
  task: Task<Ctx>,
) => any | Promise<any>;

export interface TaskMetadata {
  depth: number;
  index: number;
  startTime: number;
  stopTime: number;
}

export default class Task<Ctx extends Context> extends Emitter {
  action: TaskAction<Ctx>;

  // @ts-ignore Set after instantiation
  context: Ctx;

  title: string;

  metadata: TaskMetadata = {
    depth: 0,
    index: 0,
    startTime: 0,
    stopTime: 0,
  };

  output: string = '';

  parent: Task<Ctx> | null = null;

  status: Status = STATUS_PENDING;

  statusText: string = '';

  tasks: Task<Ctx>[] = [];

  constructor(title: string, action: TaskAction<Ctx>) {
    super();

    if (!title || typeof title !== 'string') {
      throw new Error('Tasks require a title.');
    }

    if (action !== null && typeof action !== 'function') {
      throw new Error('Tasks require an executable function.');
    }

    this.action = action;
    this.status = STATUS_PENDING;
    this.title = title;
  }

  /**
   * Return true if the task failed when executing.
   */
  hasFailed(): boolean {
    return this.status === STATUS_FAILED;
  }

  /**
   * Return true if the task executed successfully.
   */
  hasPassed(): boolean {
    return this.status === STATUS_PASSED;
  }

  /**
   * Return true if the task has been completed in any form.
   */
  isComplete(): boolean {
    return this.hasPassed() || this.hasFailed() || this.isSkipped();
  }

  /**
   * Return true if the task has not been executed yet.
   */
  isPending(): boolean {
    return this.status === STATUS_PENDING;
  }

  /**
   * Return true if the task is currently running.
   */
  isRunning(): boolean {
    return this.status === STATUS_RUNNING;
  }

  /**
   * Return true if the task was or will be skipped.
   */
  isSkipped(): boolean {
    return this.status === STATUS_SKIPPED;
  }

  /**
   * Run the current task by executing it and performing any before and after processes.
   */
  async run(context: Ctx, value: any): Promise<any> {
    this.setContext(context);
    this.emit('run', [value]);

    if (this.isSkipped()) {
      this.status = STATUS_SKIPPED;
      this.emit('skip', [value]);

      return Promise.resolve(value);
    }

    this.status = STATUS_RUNNING;
    this.metadata.startTime = Date.now();

    try {
      this.output = await this.action(context, value, this);
      this.status = STATUS_PASSED;
      this.metadata.stopTime = Date.now();
      this.emit('pass', [this.output]);
    } catch (error) {
      this.status = STATUS_FAILED;
      this.metadata.stopTime = Date.now();
      this.emit('fail', [error]);

      throw error;
    }

    this.statusText = '';

    return this.output;
  }

  /**
   * Set the context to be passed around.
   */
  setContext(context: Ctx): this {
    this.context = context;

    return this;
  }

  /**
   * Mark a task as skipped if the condition is true.
   */
  skip(condition: boolean = true): this {
    if (condition) {
      this.status = STATUS_SKIPPED;
    }

    return this;
  }
}
