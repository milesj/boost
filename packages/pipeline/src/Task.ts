import { Blueprint, Predicates } from '@boost/common';
import WorkUnit from './WorkUnit';

export default class Task<Input = unknown, Output = Input> extends WorkUnit<{}, Input, Output> {
	// A task is simply a work unit that executes a standard function.
	// It doesn't need configurable options, so implement an empty blueprint.
	blueprint(preds: Predicates): Blueprint<object> {
		return {};
	}
}
