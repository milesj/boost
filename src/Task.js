/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import Promise from 'bluebird';
import { frames } from 'elegant-spinner';
import { PENDING, RUNNING, SKIPPED, PASSED, FAILED } from './constants';

import type { Config, Result, ResultPromise, Status, TaskCallback } from './types';

export default class Task {
  action: ?TaskCallback = null;
  config: Config = {};
  context: Object = {};
  frame: number = 0;
  title: string = '';
  status: Status = PENDING;
  subroutines: Task[] = [];
  subtasks: Task[] = [];

  constructor(title: string, action: ?TaskCallback = null, defaultConfig: Config = {}) {
    if (!title || typeof title !== 'string') {
      throw new Error('Tasks require a title.');
    }

    if (action !== null && typeof action !== 'function') {
      throw new Error('Tasks require an executable function.');
    }

    this.action = action;
    this.config = { ...defaultConfig };
    this.status = action ? PENDING : SKIPPED;
    this.title = title;
  }

  /**
   * Return true if the task failed when executing.
   */
  hasFailed(): boolean {
    return (this.status === FAILED);
  }

  /**
   * Return true if the task executed successfully.
   */
  hasPassed(): boolean {
    return (this.status === PASSED);
  }

  /**
   * Return true if the task has not been executed yet.
   */
  isPending(): boolean {
    return (this.status === PENDING);
  }

  /**
   * Return true if the task is currently running.
   */
  isRunning(): boolean {
    return (this.status === RUNNING);
  }

  /**
   * Return true if the task was or will be skipped.
   */
  isSkipped(): boolean {
    return (this.status === SKIPPED);
  }

  /**
   * Run the current task by executing it and performing any
   * before and after processes.
   */
  run(value: Result, context: Object = {}): ResultPromise {
    this.context = context;

    if (this.isSkipped() || !this.action) {
      return Promise.resolve(value);
    }

    this.status = RUNNING;

    return new Promise((resolve) => {
      // $FlowIgnore We check if nullable above
      resolve(this.action(value, context));
    }).then(
      (result) => {
        this.status = PASSED;

        return result;
      },
      (error) => {
        this.status = FAILED;

        throw error;
      },
    );
  }

  /**
   * Mark a task as skipped if the condition is true.
   */
  skip(condition: boolean = true): this {
    if (condition) {
      this.status = SKIPPED;
    }

    return this;
  }

  /**
   * Create a spinner and update the frames each call.
   */
  spinner(): string {
    // eslint-disable-next-line no-plusplus
    this.frame = ++this.frame % frames.length;

    return frames[this.frame];
  }

  /**
   * Wrap a value in a promise if it has not already been.
   */
  wrap(value: Result): ResultPromise {
    return (value instanceof Promise) ? value : Promise.resolve(value);
  }
}
