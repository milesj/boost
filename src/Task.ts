/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { frames } from 'elegant-spinner';
import { Struct } from 'optimal';
import {
  STATUS_PENDING,
  STATUS_RUNNING,
  STATUS_SKIPPED,
  STATUS_PASSED,
  STATUS_FAILED,
} from './constants';
import { Context, Status } from './types';

export interface TaskInterface {
  status: Status;
  statusText: string;
  subroutines: TaskInterface[];
  subtasks: TaskInterface[];
  title: string;
  isPending(): boolean;
  isRunning(): boolean;
  isSkipped(): boolean;
  hasFailed(): boolean;
  hasPassed(): boolean;
  run<T>(context: any, initialValue?: T | null): Promise<any>;
  skip(condition?: boolean): this;
  spinner(): string;
}

export type TaskAction<Tx extends Context> = (context: Tx, value: any) => any | Promise<any>;

export default class Task<To extends Struct, Tx extends Context> implements TaskInterface {
  action: TaskAction<Tx> | null = null;

  // @ts-ignore Set after instantiation
  context: Tx;

  frame: number = 0;

  options: To;

  title: string = '';

  status: Status = STATUS_PENDING;

  statusText: string = '';

  subroutines: TaskInterface[] = [];

  subtasks: TaskInterface[] = [];

  constructor(title: string, action: TaskAction<Tx> | null = null, options: Partial<To> = {}) {
    if (!title || typeof title !== 'string') {
      throw new Error('Tasks require a title.');
    }

    if (action !== null && typeof action !== 'function') {
      throw new Error('Tasks require an executable function.');
    }

    this.action = action;
    this.options = {
      // @ts-ignore
      ...options,
    };
    this.status = action ? STATUS_PENDING : STATUS_SKIPPED;
    this.title = title;
  }

  /**
   * Return true if the task failed when executing.
   */
  hasFailed(): boolean {
    return this.status === STATUS_FAILED;
  }

  /**
   * Return true if the task executed successfully.
   */
  hasPassed(): boolean {
    return this.status === STATUS_PASSED;
  }

  /**
   * Return true if the task has not been executed yet.
   */
  isPending(): boolean {
    return this.status === STATUS_PENDING;
  }

  /**
   * Return true if the task is currently running.
   */
  isRunning(): boolean {
    return this.status === STATUS_RUNNING;
  }

  /**
   * Return true if the task was or will be skipped.
   */
  isSkipped(): boolean {
    return this.status === STATUS_SKIPPED;
  }

  /**
   * Run the current task by executing it and performing any before and after processes.
   */
  run<T>(context: Tx, initialValue: T | null = null): Promise<any> {
    // Don't spread context as to preserve references
    this.context = context;

    if (this.isSkipped() || !this.action) {
      this.status = STATUS_SKIPPED;

      return Promise.resolve(initialValue);
    }

    this.status = STATUS_RUNNING;

    return (
      Promise.resolve(initialValue)
        // @ts-ignore
        .then(value => this.action(context, value))
        .then(
          result => {
            this.status = STATUS_PASSED;
            this.statusText = '';

            return result;
          },
          error => {
            this.status = STATUS_FAILED;

            throw error;
          },
        )
    );
  }

  /**
   * Mark a task as skipped if the condition is true.
   */
  skip(condition: boolean = true): this {
    if (condition) {
      this.status = STATUS_SKIPPED;
    }

    return this;
  }

  /**
   * Create a spinner and update the frames each call.
   */
  spinner(): string {
    // eslint-disable-next-line no-plusplus
    this.frame = ++this.frame % frames.length;

    return frames[this.frame];
  }

  /**
   * Wrap a value in a promise if it has not already been.
   */
  wrap<T>(value: T): Promise<any> {
    return value instanceof Promise ? value : Promise.resolve(value);
  }
}
