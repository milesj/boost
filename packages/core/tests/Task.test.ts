import Task from '../src/Task';
import {
  STATUS_PENDING,
  STATUS_RUNNING,
  STATUS_SKIPPED,
  STATUS_PASSED,
  STATUS_FAILED,
} from '../src/constants';

describe('Task', () => {
  let task: Task<any>;

  beforeEach(() => {
    task = new Task('title', (con, value) => value * 2);
  });

  describe('constructor()', () => {
    it('errors if no title', () => {
      expect(() => new Task('')).toThrowErrorMatchingSnapshot();
    });

    it('errors if title is not a string', () => {
      // @ts-ignore
      expect(() => new Task(123)).toThrowErrorMatchingSnapshot();
    });

    it('errors if action is not a function', () => {
      // @ts-ignore
      expect(() => new Task('title', 123)).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if action is null', () => {
      expect(() => new Task('title')).not.toThrow();
    });

    it('marks the task as skipped if no action', () => {
      task = new Task('title');

      expect(task.isSkipped()).toBe(true);
    });
  });

  describe('hasFailed()', () => {
    it('returns a boolean for STATUS_FAILED status state', () => {
      expect(task.hasFailed()).toBe(false);

      task.status = STATUS_FAILED;

      expect(task.hasFailed()).toBe(true);
    });
  });

  describe('hasPassed()', () => {
    it('returns a boolean for STATUS_PASSED status state', () => {
      expect(task.hasPassed()).toBe(false);

      task.status = STATUS_PASSED;

      expect(task.hasPassed()).toBe(true);
    });
  });

  describe('isPending()', () => {
    it('returns a boolean for STATUS_PENDING status state', () => {
      expect(task.isPending()).toBe(true);

      task.status = STATUS_PASSED;

      expect(task.isPending()).toBe(false);
    });
  });

  describe('isRunning()', () => {
    it('returns a boolean for STATUS_RUNNING status state', () => {
      expect(task.isRunning()).toBe(false);

      task.status = STATUS_RUNNING;

      expect(task.isRunning()).toBe(true);
    });
  });

  describe('isSkipped()', () => {
    it('returns a boolean for STATUS_SKIPPED status state', () => {
      expect(task.isSkipped()).toBe(false);

      task.status = STATUS_SKIPPED;

      expect(task.isSkipped()).toBe(true);
    });
  });

  describe('run()', () => {
    it('resolves a value with the action', async () => {
      try {
        expect(await task.run({}, 123)).toBe(246);
        expect(task.status).toBe(STATUS_PASSED);
      } catch (error) {
        expect(true).toBe(false); // Would fail
      }
    });

    it('resolves a value if the task should be skipped', async () => {
      try {
        task.status = STATUS_SKIPPED;

        expect(await task.run({}, 123)).toBe(123);
      } catch (error) {
        expect(true).toBe(false); // Would fail
      }
    });

    it('rejects the value if the action throws an error', async () => {
      try {
        task.action = () => {
          throw new Error('Oops');
        };

        await task.run({}, 123);

        expect(true).toBe(false); // Would fail
      } catch (error) {
        expect(error).toEqual(new Error('Oops'));
        expect(task.status).toBe(STATUS_FAILED);
      }
    });

    it('passes the value through when no action exists', async () => {
      task.action = null;

      expect(await task.run({}, 123)).toBe(123);
    });

    it('passes a context to the action', async () => {
      const context = { count: 1 };

      /* eslint-disable no-param-reassign */
      task.action = (con, value) => {
        con.count += 1;
        con.foo = 'bar';
      };
      /* eslint-enable */

      await task.run(context, 123);

      expect(context).toEqual({
        count: 2,
        foo: 'bar',
      });
    });

    it('sets times on success', async () => {
      await task.run({}, 123);

      expect(task.startTime).not.toBe(0);
      expect(task.stopTime).not.toBe(0);
    });

    it('sets times on failure', async () => {
      try {
        task.action = () => {
          throw new Error('Oops');
        };

        await task.run({}, 123);
      } catch (error) {
        // Skip
      }

      expect(task.startTime).not.toBe(0);
      expect(task.stopTime).not.toBe(0);
    });

    it('passes task as 3rd argument to action', async () => {
      const spy = jest.fn();

      task.action = spy;

      await task.run({}, 123);

      expect(spy).toHaveBeenCalledWith({}, 123, task);
    });
  });

  describe('setContext()', () => {
    it('sets the context', () => {
      task.setContext({ foo: 'bar' });

      expect(task.context).toEqual({ foo: 'bar' });
    });
  });

  describe('skip()', () => {
    it('marks a task as STATUS_SKIPPED', () => {
      expect(task.status).toBe(STATUS_PENDING);

      task.skip();

      expect(task.status).toBe(STATUS_SKIPPED);
    });

    it('evaluates a condition to determine whether to skip', () => {
      expect(task.status).toBe(STATUS_PENDING);

      // @ts-ignore
      task.skip(1 === 2);

      expect(task.status).toBe(STATUS_PENDING);
    });
  });
});
