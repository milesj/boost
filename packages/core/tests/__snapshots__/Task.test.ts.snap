// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task constructor() errors if action is not a function 1`] = `"Tasks require an executable function."`;

exports[`Task constructor() errors if no title 1`] = `"Tasks require a title."`;

exports[`Task constructor() errors if title is not a string 1`] = `"Tasks require a title."`;
